p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+/*
+ * Hashes
+ */
+
+/// Supported hash algorithms
+enum HashAlgorithm {
+  crc32,
+  crc32_custom,
+  crc16,
+  crc16_custom,
+  random,           /// are random hash algorithms useful?
+  identity
+}
+
+/// Hash function
+///
+/// @param Algo The algorithm to use for computation (@see HashAlgorithm).
+/// @param O    The type of the return value of the hash.
+///
+/// Example usage:
+/// parser P() {
+///   Hash<crc16, bit<56>> h();
+///   bit<56> hash_value = h.getHash(16, buffer, 100);
+///   ...
+/// }
+extern Hash<Algo, O> {
+  /// Constructor
+  Hash();
+
+  /// compute the hash for data
+  O getHash<T, D, M>(in T base, in D data, in M max);
+
+  /*
+  @ControlPlaneAPI
+  { }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/// Checksum computation
+///
+/// @param W    The width of the checksum
+extern Checksum<W> {
+  Checksum(HashAlgorithm hash);          /// constructor
+  void clear();              /// prepare unit for computation
+  void update<T>(in T data); /// add data to checksum
+  void remove<T>(in T data); /// remove data from existing checksum
+  W    get();      	     /// get the checksum for data added since last clear
+
+  /*
+  @ControlPlaneAPI
+  { }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Counters
+ *
+ * Counters are a simple mechanism for keeping statistics about the
+ * packets that trigger a table in a Match Action unit.
+ *
+ * Direct counters fire when the count method is ivoked in an action, and
+ * have an instance for each entry in the table.
+ */
+
+/// Counter types
+enum CounterType_t {
+    packets,
+    bytes,
+    packets_and_bytes
+}
+
+/// Counter: RFC-XXXX
+extern Counter<W, S> {
+  Counter(S n_counters, W size_in_bits, CounterType_t counter_type);
+  void count(in S index, in W increment);
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in S index);
+    W    sync_read<W> (in S index);
+    void set          (in S index, in W seed);
+    void reset        (in S index);
+    void start        (in S index);
+    void stop         (in S index);
+  }
+  */
+}
+
+/// DirectCounter: RFC-XXXX
+extern DirectCounter<W> {
+  DirectCounter(entry_key key, CounterType_t counter_type);
+  void count();
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in entry_key key);
+    W    sync_read<W> (in entry_key key);
+    void set          (in W seed);
+    void reset        (in entry_key key);
+    void start        (in entry_key key);
+    void stop         (in entry_key key);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Meters: RFC 2698
+ *
+ * Meters are a more complex mechanism for keeping statistics about
+ * the packets that trigger a table. The meters specified in the PSA
+ * are 3-color meters.
+ */
+
+/// Meter types
+enum MeterType_t {
+    packets,
+    bytes
+}
+/// Meter colors
+enum MeterColor_t { RED, GREEN, YELLOW };
```

Comment:
@vgurevich I have never heard of a use case for storing this red/green/yellow value in the packet.  Have you?  Mostly I have seen a 3-way branch in the code after updating the meter, that does different things depending upon the value of red/green/yellow, e.g. drop on red, modify DSCP on yellow, no-op on green.  That said, I have no problem using types that are easily converted to/from bit vectors.